## Example: Building an Intensity Matrix

First, setup the paths to the datafiles and the output directory, then import JCAMP_reader.

In [1]:
import pathlib
data_directory = pathlib.Path(".").resolve().parent.parent / "pyms-data"
# Change this if the data files are stored in a different location

output_directory = pathlib.Path(".").resolve() / "output"

from pyms.GCMS.IO.JCAMP import JCAMP_reader

Read the raw data files.

In [2]:
jcamp_file = data_directory / "gc01_0812_066.jdx"
data = JCAMP_reader(jcamp_file)
data

 -> Reading JCAMP file '/home/runner/work/PyMassSpec/PyMassSpec/pyms-data/gc01_0812_066.jdx'


<GCMS_data(305.582 - 4007.722 seconds, time step 0.3753183292781833, 9865 scans)>

Then the data can be converted to an |IntensityMatrix| using the function 
|build_intensity_matrix()| from |pyms.IntensityMatrix|.

The default operation of |build_intensity_matrix()| is to use a bin interval of
 one and treat the masses as floating point numbers. The default intensity 
 matrix can be built as follows:

In [3]:
from pyms.IntensityMatrix import build_intensity_matrix

im = build_intensity_matrix(data)

im

The size as the number of scans and the number of bins can be returned with:

In [4]:
im.size

(9865, 551)

There are 9865 scans and 551 bins in this example.

The raw masses have been binned into new mass units based on the minimum mass
in the raw data and the bin size. A list of the first ten new masses can be 
obtained as follows:

In [5]:
im.mass_list[:10]

[50.0, 51.0, 52.0, 53.0, 54.0, 55.0, 56.0, 57.0, 58.0, 59.0]

The attributes |im.min_mass| and |im.max_mass| return the minimum and maximum mass:

In [6]:
im.min_mass

50.0

In [7]:
im.max_mass

600.0

It is also possible to search for a particular mass, by finding the index of
the binned mass closest to the desired mass. For example, the index of the
closest binned mass to a mass of 73.3 |m/z| can be found by using the
methods |im.get_index_of_mass()|:

In [8]:
index = im.get_index_of_mass(73.3)

index

23

The value of the closest mass can be returned by the method
|im.get_mass_at_index()|:

In [9]:
im.get_mass_at_index(index)

73.0

A mass of 73.0 is returned in this example.

## Build intensity matrix parameters

The bin interval can be set to values other than one, and binning boundaries
can also be adjusted. In the example below, to fit the 0.5 bin interval, the
upper and lower boundaries are set to ± 0.25.

In [10]:
im = build_intensity_matrix(data, 0.5, 0.25, 0.25)

im

The size of the intensity matrix will reflect the change in the number of bins:

In [11]:
im.size

(9865, 1101)

In [12]:
im.mass_list[:10]

[50.0, 50.5, 51.0, 51.5, 52.0, 52.5, 53.0, 53.5, 54.0, 54.5]

In this example there are 9865 scans (as before), but 1101 bins.

The index and binned mass of the mass closest to 73.3 should also reflect the
different binning.

In [13]:
index = im.get_index_of_mass(73.3)

index

47

In [14]:
im.get_mass_at_index(index)

73.5

## Build integer mass intensity matrix

It is also possible to build an intensity matrix with integer masses and a bin
interval of one using |build_intensity_matrix_i()|. The default range for the 
binning is -0.3 and +0.7 mass units. The function is imported from 
|pyms.IntensityMatrix|:

In [15]:
from pyms.IntensityMatrix import build_intensity_matrix_i

im = build_intensity_matrix_i(data)
im

In [16]:
im.size

(9865, 551)

In [17]:
im.mass_list[:10]

[50, 51, 52, 53, 54, 55, 56, 57, 58, 59]

The masses are now integers.

In [18]:
index = im.get_index_of_mass(73.3)
index

23

In [19]:
im.get_mass_at_index(index)

73

The lower and upper bounds can be adjusted with |build_intensity_matrix_i(data, lower, upper)|.